qa/workunits/rbd: cover native migration between namespaces

sunilangadi2 opened the pull request:
[RBD] : Added tests related to migration with namespace

-  Migration of images from namespace to namespace 
-  Migration of images from namespace to non-namespace
-  Migration of images from non-namespace to namespace


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

idryomov commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -569,6 +629,7 @@ EOF
     kill_nbd_server
 }
 
+
```
stray whitespace
```suggestion
```

sunilangadi2 replied:
sorry missed it, updated.

idryomov commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -39,6 +41,23 @@ create_base_image() {
     rbd bench --io-type write --io-pattern rand --io-size=4K --io-total 128M ${image}
 }
 
+create_base_image_with_namespace() {
```
This function is a copy of `create_base_image()` with the exception of `rbd namespace create` command.  I'd suggest creating the namespace at the top level and passing it to the existing `create_base_image()` as part of the image spec:
```
rbd namespace create rbd/${NAMESPACE1}
create_base_image rbd/${NAMESPACE1}/${IMAGE1}
test_import_native_format_with_namespace ...
```

idryomov commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -39,6 +41,23 @@ create_base_image() {
+    local image=$2
+
+    # Create a new namespace and new image within that namespace
+    rbd namespace create rbd/${namespace}
```
Both namespaces need to be removed in `cleanup()` -- it should be possible to run the script multiple times in a row without any manual cleanup in between.

sunilangadi2 replied:
added `namespace` removal in in cleanup()

idryomov commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -199,6 +218,47 @@ EOF
 }
 
+# Function to test the import of native format with namespace
+test_import_native_format_with_namespace() {
```
Currently this function is covering "Migration from namespace to non-namespace" and "Migration from namespace to namespace" scenarios.  Let's cover "Migration from non-namespace to namespace" as well for completeness.

sunilangadi2 replied:
covered the suggested scenario as well.

idryomov commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -199,6 +218,47 @@ EOF
+    compare_images "rbd/${base_namespace}/${base_image}@2" "${dest_image}@2"
+
+    # Migration from namespace to namespace
+    rbd namespace create rbd/${dest_namespace}
```
I'd suggest creating this namespace at the top level as well, together with `rbd/${NAMESPACE1}`.

sunilangadi2 replied:
Done

sunilangadi2 commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -6,6 +6,8 @@ IMAGE1=image1
 IMAGE3=image3
 IMAGES="${IMAGE1} ${IMAGE2} ${IMAGE3}"
+NAMESPACE1=test_namespace1
+NAMESPACE2=test_namespace2
```
Done

sunilangadi2 commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -587,4 +648,6 @@ test_import_nbd_stream_qcow2 ${IMAGE2} ${IMAGE3}
 
+create_base_image_with_namespace ${NAMESPACE1} ${IMAGE1}
+test_import_native_format_with_namespace ${NAMESPACE1} ${IMAGE1} ${NAMESPACE2} ${IMAGE2}
 echo OK
```
Done

sunilangadi2 commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -199,6 +218,47 @@ EOF
     remove_image "${dest_image}"
 }
 
+# Function to test the import of native format with namespace
```
Done

sunilangadi2 commented:
Thanks @idryomov for the review,
updated the code by addressing all review comments,
Please review it again.

sunilangadi2 commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -199,6 +213,58 @@ EOF
+        rbd migration prepare --import-only --source-spec-path ${TEMPDIR}/spec.json ${dest_image}
+        rbd migration execute ${dest_image}
+        rbd migration commit ${dest_image}
+        compare_images "${source_image}@1" "${dest_image}@1"
```
https://github.com/ceph/ceph/blob/main/qa/workunits/rbd/cli_migration.sh#L115
here as migration involves with second snapshot,
what is the point of comparing with first snapshot
here https://github.com/ceph/ceph/blob/main/qa/workunits/rbd/cli_migration.sh#L123?

sunilangadi2 commented:
jenkins test make check

nbalacha commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -69,6 +72,17 @@ remove_images() {
+}
+
+remove_namespaces() {
+    rbd namespace remove rbd/${NAMESPACE1} || true
```
Maybe just call remove_namespace()  for each namespace ?

sunilangadi2 commented:
@idryomov @nbalacha addressed review comments,
also tested the script locally test passed
https://privatebin.corp.redhat.com/?d0e901af75828647#C7FXxDwe7TZHuydzfuZSD9fQJeRrqGGxhbH2DpWNhE6U
please review it again.

sunilangadi2 commented:
jenkins test make check

sunilangadi2 commented:
jenkins test api

sunilangadi2 commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -40,10 +44,12 @@ create_base_image() {
-    local image=$1
+    local image_path=$1
+    # If image is from namespace
+    local image=$(basename "$image_path")
```
ok thanks for that logic, implemented the same way.

sunilangadi2 commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -84,6 +108,25 @@ show_diff()
+    local dest_image=$(basename "$destination")
+    local ret=0
+
+    export_raw_image "${source}"
```
Now removed `compare_namespace_images()` but `source image` also needs to be exported very time as we doing different scenarios with namespace based images

idryomov replied:
> `source image` also needs to be exported very time as we doing different scenarios with namespace based images

Sorry, I'm not following.  The source image doesn't change, so calling `export_base_image rbd/${NAMESPACE1}/${IMAGE1}` after `create_base_image rbd/${NAMESPACE1}/${IMAGE1}` once should be sufficient.

sunilangadi2 commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -199,6 +242,61 @@ EOF
+{
+  "type": "native",
+  "pool_id": ${pool_id},
+  "pool_namespace": "${namespace}",
```
> Have you considered adapting test_import_native_format() to work with namespace-containing image specs? 

tried but as it involves with different scenario instead of messing with it created separate function,
but after your suggestion on how to call same function for namespace based images
I thought this would be much easier than earlier with some tweaks to existing,
Thanks for the logic!
now implemented that way.

sunilangadi2 commented:
Tested the changes locally, all tests got passed
https://privatebin.corp.redhat.com/?24234c7ac194011a#5U4WCm7hxuXXmedCk2rkYEixNBjuWttSB4PX3V7pF6a7
@idryomov Please review it.

sunilangadi2 commented:
jenkins test make check

idryomov commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -102,23 +135,32 @@ test_import_native_format() {
+    fi
+
+    # Execute only if it's a non-namespace based image
+    if [[ -z "$namespace" ]]; then
```
Is this `if` condition really needed?  Can `rbd/${base_image}@2` argument to `rbd migration prepare` command be replaced with just `${base_image}@2` so that this test is performed in all configurations?  It wouldn't change the existing logic at all -- the pool is specified redundantly.

idryomov commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -102,23 +135,32 @@ test_import_native_format() {
-  "pool_namespace": "",
-  "image_name": "${base_image}",
+  "pool_namespace": "${namespace}",
+  "image_name": "$(basename "$base_image")",
```
I'd suggest renaming the existing `base_image` and `dst_image` variables to `base_image_spec` and `dst_image_spec` respectively, the new `namespace` to `base_namespace` and having `base_image` contain just the image name so that `basename` doesn't need to be called repeatedly.

sunilangadi2 replied:
Done as per the suggestion.

idryomov commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -134,67 +176,69 @@ EOF
-
-    compare_images "${base_image}@1" "${dest_image}@1"
-    compare_images "${base_image}@2" "${dest_image}@2"
+    if [[ -z "$namespace" ]]; then
```
I don't think this `if` condition is needed either.  The point of
> * add `"pool_namespace": "${namespace}"` to the source spec option value everywhere it's passed with `--source-spec`

step that I suggested in https://github.com/ceph/ceph/pull/60042#discussion_r1804671256 was to have the same set of tests executed in all configurations.  Disabling the source spec validity tests in two out of four configurations isn't ideal.

sunilangadi2 replied:
ok initially i thought those tests only for pool-based images, now added  `"pool_namespace": "${namespace}"` to the source spec option.

idryomov commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -89,7 +112,17 @@ compare_images() {
     local ret=0
 
-    export_raw_image ${dst_image}
+   # Export both if image is from a namespace
```
There is another thread on this, commenting here just in case: I'm still not convinced that `compare_images()` should ever (re-)export the source image.

idryomov commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -69,6 +76,22 @@ remove_images() {
 
+remove_namespace() {
+    local namespace=$1
+    local images=$(rbd ls --namespace ${namespace})
+    for IMAGE in ${images}; do
+        remove_image rbd/${namespace}/${IMAGE} || true
```
`remove_images()` works on a hard-coded list of image names, I'd suggest to do the same here:
```suggestion
    for image in ${IMAGES}; do
        remove_image rbd/${namespace}/${image} || true
```

sunilangadi2 replied:
But all the list of images from `${IMAGES}` may not belongs to that `namespace1` or `namespace2`
hence I listing the images inside namespace then removing it.

idryomov replied:
That is fine -- this is a cleanup routine which is explicitly written to not fail if an image isn't there, or if it's there but not being migrated, etc.  Note the use of `|| true`.

sunilangadi2 replied:
ok got it thanks, now made the changes accordingly.

sunilangadi2 commented:
@idryomov thanks for the solutions, now addressed all comments as per the suggestion
all tests got passed
https://privatebin.corp.redhat.com/?d0fe6068e8e2e1c3#B1fQR6jmTaAjyM3MVg8F6jCNAWPLxr4r16NGHJQ6fw6X
Please review it.

sunilangadi2 commented:
jenkins test api

idryomov commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -69,6 +76,21 @@ remove_images() {
     done
 }
 
+remove_namespace() {
```
It occurs to me that we could get rid of this function by defining
```
IMAGES="${IMAGE1} ${IMAGE2} ${IMAGE3} ${NAMESPACE1}/${IMAGE1} ${NAMESPACE2}/${IMAGE2}"
```
and changing `remove_namespaces()` definition to:
```
remove_namespaces() {
    for namespace in ${NAMESPACES}
    do
        rbd namespace remove rbd/${namespace} || true
    done
}
```

sunilangadi2 replied:
Thanks for this logic, now made the changes accordingly.

idryomov commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -90,6 +112,14 @@ compare_images() {
 
     export_raw_image ${dst_image}
+
+    # Check if either src_image or dst_image is from a namespace
+    if [[ "$src_image" == rbd/*/* || "$dst_image" == rbd/*/* ]]; then
+       # Replace slashes (/) with underscores (_) for namespace images
+       src_image="${src_image//\//_}"
+       dst_image="${dst_image//\//_}"
+    fi
```
This `if` statement seems redundant.  In case there are no slashes, `${VAR//\//_}` construct would do nothing:
```suggestion
    # Replace slashes (/) with underscores (_) for namespace images
    src_image="${src_image//\//_}"
    dst_image="${dst_image//\//_}"
```

idryomov commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -99,104 +129,112 @@ compare_images() {
-    rbd migration prepare --import-only "rbd/${base_image}@2" ${dest_image}
-    rbd migration abort ${dest_image}
+    # if image is from namespace
+    local namespace=""
```
For consistency with `base_image_spec` and `base_image` variables:
```suggestion
    local base_namespace=""
```

sunilangadi2 replied:
updated the variables as per the suggestion.

idryomov commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -99,104 +129,112 @@ compare_images() {
+    if [[ "$base_image_spec" == rbd/*/* ]]; then
+        namespace=$(basename "$(dirname "$base_image_spec")")
+    fi
+    base_image=$(basename "$base_image_spec")
```
I'd move this assignment into the `if` branch for clarity:
```
    # if base image is from namespace
    local base_namespace=""
    local base_image=${base_image_spec}
    if [[ "${base_image_spec}" == rbd/*/* ]]; then
        base_namespace=$(basename "$(dirname "${base_image_spec}")")
        base_image=$(basename "${base_image_spec}")
    fi
```

sunilangadi2 replied:
thanks for the Example code, now updated the code as per the suggestion.

idryomov commented on qa/workunits/rbd/cli_migration.sh:
```diff
@@ -99,104 +129,112 @@ compare_images() {
 
-    compare_images "${base_image}@1" "${dest_image}@1"
-    compare_images "${base_image}@2" "${dest_image}@2"
+    rbd migration execute ${dest_image_spec}
```
I think you accidentally added a blank line before `rbd migration execute` line that wasn't there and it confuses GitHub's diff highlighting algorithm here.

sunilangadi2 replied:
ok now i removed it.

idryomov reviewed (commented):
Overall LGTM, just a few nits remaining.

idryomov commented:
https://pulpito.ceph.com/dis-2024-10-24_07:40:53-rbd-main-distro-default-smithi/
